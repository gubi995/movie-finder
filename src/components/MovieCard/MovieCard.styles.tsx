import styled from '@emotion/styled';
import { Card, CardContent } from '@mui/material';

import { BACKGROUND } from '@/styles/linear-gradients';

export const TopSection = styled.div`
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
`;

export const TitleAndDateContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
`;

export const Score = styled.div`
  display: grid;
  place-items: center;
  padding: 0.5em;
  border-radius: 50%;
  width: 30px;
  height: 30px;
  font-weight: bold;
  background: ${BACKGROUND};
  color: white;
`;

export const StyledCardContent = styled(CardContent)`
  display: flex;
  flex-direction: column;
  padding: 2rem;
  gap: 1rem;
  flex: 1 1;
`;

export const ChipContainer = styled.div`
  display: flex;
  gap: 0.5rem;
`;

export const StyledCard = styled(Card)`
  display: flex;
  flex-direction: column;

  @media (min-width: 768px) {
    flex-direction: row;
  }
`;
