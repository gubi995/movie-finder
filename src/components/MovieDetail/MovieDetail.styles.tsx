import styled from '@emotion/styled';
import { Typography } from '@mui/material';

export const MovieDetailContent = styled.div`
  display: flex;
  flex-direction: column;
  gap: 1rem;
  overflow: hidden;
`;

export const LinkContainer = styled.div`
  display: flex;
  gap: 1rem;
  align-items: center;
`;

export const LinksLabel = styled(Typography)`
  text-transform: uppercase;
  font-size: 0.8em;
`;

export const Detail = styled(Typography)`
  overflow: auto;
`;
