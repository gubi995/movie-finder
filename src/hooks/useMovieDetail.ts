import { useQuery } from 'react-query';

import { MovieDetail } from '@/model/movie';
import { fetchWikipediaPage } from '@/api/wikipedia';

export const useMovieDetail = (title: string) =>
  useQuery<MovieDetail, Error>(['movieDetail', title], () => fetchWikipediaPage(title), {
    refetchOnWindowFocus: false,
  });
