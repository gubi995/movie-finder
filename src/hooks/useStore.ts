import { ReactNode } from 'react';
import create from 'zustand';

interface State {
  searchTerm: string;
  modalContent: ReactNode;
  updateSearchTerm: (term: string) => void;
  setModalContent: (term: ReactNode) => void;
}

export const useStore = create<State>((set) => ({
  searchTerm: '',
  modalContent: null,
  updateSearchTerm: (term: string) => set({ searchTerm: term }),
  setModalContent: (content: ReactNode) => set({ modalContent: content }),
}));
