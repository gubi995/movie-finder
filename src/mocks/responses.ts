export const MOCK_GRAPHQL_RESPONSE = {
  searchMovies: [
    {
      id: '550',
      name: 'Fight Club',
      genres: [
        {
          name: 'Drama',
        },
      ],
      score: 8.4,
      overview:
        'A ticking-time-bomb insomniac and a slippery soap salesman channel primal male aggression into a shocking new form of therapy. Their concept catches on, with underground "fight clubs" forming in every town, until an eccentric gets in the way and ignites an out-of-control spiral toward oblivion.',
      releaseDate: '1999-10-15T00:00:00.000Z',
    },
    {
      id: '629017',
      name: 'Run Hide Fight',
      genres: [
        {
          name: 'Crime',
        },
        {
          name: 'Thriller',
        },
      ],
      score: 6.6,
      overview:
        'A 17-year-old girl uses her wits, survival skills, and compassion to fight for her life, and those of her fellow classmates, against a group of live-streaming school shooters.',
      releaseDate: '2020-09-10T00:00:00.000Z',
    },
  ],
};

export const MOCK_WIKIPEDIA_RESPONSE = {
  batchcomplete: '',
  query: {
    pages: {
      '1009041': {
        pageid: 1009041,
        title: 'Fight Club',
        extlinks: [{ '*': 'https://www.imdb.com/title/tt0137523/' }],
        extract:
          'Fight Club is a 1999 American film directed by David Fincher and starring Brad Pitt, Edward Norton, and Helena Bonham Carter. It is based on the 1996 novel of the same name by Chuck Palahniuk. Norton plays the unnamed narrator, who is discontented with his white-collar job. He forms a "fight club" with soap salesman Tyler Durden (Pitt), and becomes embroiled in a relationship with a destitute woman, Marla Singer (Bonham Carter).\nPalahniuk\'s novel was optioned by Fox 2000 Pictures producer Laura Ziskin, who hired Jim Uhls to write the film adaptation.',
      },
    },
  },
};
