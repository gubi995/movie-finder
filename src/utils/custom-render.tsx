import { render, RenderOptions } from '@testing-library/react';
import { FC, ReactElement } from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});

const AllTheProviders: FC = ({ children }) => {
  return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>;
};

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';

export { queryClient };

export { customRender as render };
