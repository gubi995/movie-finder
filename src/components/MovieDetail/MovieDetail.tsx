import { Chip, Skeleton, Typography } from '@mui/material';
import { VFC } from 'react';

import { useMovieDetail } from '@/hooks/useMovieDetail';

import { LinkContainer, LinksLabel, MovieDetailContent, Detail } from './MovieDetail.styles';

interface MovieDetailProps {
  movieTitle: string;
}

const MovieDetail: VFC<MovieDetailProps> = ({ movieTitle }) => {
  const { data, isLoading, error } = useMovieDetail(movieTitle);
  const { movieDetail, links } = data ?? {};
  const { imdb, wikipedia } = links ?? {};

  const movieDescription = isLoading ? (
    <Skeleton variant="rectangular" width={700} height={200} role="progressbar" />
  ) : (
    <Detail>{movieDetail}</Detail>
  );

  const movieLinks = isLoading ? (
    <Skeleton variant="rectangular" width={200} height={20} role="progressbar" />
  ) : (
    <LinkContainer>
      <LinksLabel fontFamily="Montserrat">Links:</LinksLabel>
      {wikipedia && (
        <Chip
          label="Wikipedia"
          component="a"
          href={wikipedia}
          variant="outlined"
          clickable
          target="_blank"
        />
      )}
      {imdb && (
        <Chip label="IMDB" component="a" href={imdb} variant="outlined" clickable target="_blank" />
      )}
    </LinkContainer>
  );

  return (
    <MovieDetailContent>
      <Typography fontFamily="Montserrat" variant="h6">
        {movieTitle}
      </Typography>
      {error ? (
        <Typography variant="subtitle1">{error.message}</Typography>
      ) : (
        <>
          {movieDescription}
          {movieLinks}
        </>
      )}
    </MovieDetailContent>
  );
};

export default MovieDetail;
