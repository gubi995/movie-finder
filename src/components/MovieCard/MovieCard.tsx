import { VFC } from 'react';
import { Typography, Chip, ButtonBase } from '@mui/material';

import { useStore } from '@/hooks/useStore';
import MovieDetail from '@/components/MovieDetail';
import { Movie } from '@/model/movie';

import {
  ChipContainer,
  Score,
  StyledCard,
  StyledCardContent,
  TitleAndDateContainer,
  TopSection,
} from './MovieCard.styles';

interface MovieCardProps {
  movie: Movie;
}

const MovieCard: VFC<MovieCardProps> = ({
  movie: { name, genres, overview, releaseDate, score },
}) => {
  const setContent = useStore((state) => state.setContent);

  const openDetails = () => {
    setContent(<MovieDetail movieTitle={name} />);
  };

  return (
    <StyledCard>
      <StyledCardContent>
        <TopSection>
          <TitleAndDateContainer>
            <ButtonBase disableRipple onClick={openDetails}>
              <Typography fontFamily="Montserrat" variant="h5">
                {name}
              </Typography>
            </ButtonBase>
            <Typography color="gray">{new Date(releaseDate).toLocaleDateString()}</Typography>
          </TitleAndDateContainer>
          <Score>{score}</Score>
        </TopSection>
        <ChipContainer>
          {genres.map(({ name }) => (
            <Chip key={name} label={name} />
          ))}
        </ChipContainer>
        <Typography>{overview}</Typography>
      </StyledCardContent>
    </StyledCard>
  );
};

export default MovieCard;
