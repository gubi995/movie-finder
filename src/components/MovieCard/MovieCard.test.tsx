import userEvent from '@testing-library/user-event';

import { screen, render } from '@/utils/custom-render';
import { MOCK_GRAPHQL_RESPONSE } from '@/mocks/responses';
import * as hooks from '@/hooks/useStore';
import MovieDetail from '@/components/MovieDetail';

import MovieCard from './MovieCard';

const setModalContentMock = jest.fn();
const [movie] = MOCK_GRAPHQL_RESPONSE.searchMovies;

jest.spyOn(hooks, 'useStore').mockReturnValue(setModalContentMock);

describe('MovieCard component', () => {
  test('should render movie data', () => {
    const { baseElement } = render(<MovieCard movie={movie} />);

    expect(baseElement).toMatchSnapshot();
  });

  test('should open modal with the proper content', () => {
    render(<MovieCard movie={movie} />);

    const movieTitleButton = screen.getByRole('button');

    userEvent.click(movieTitleButton);

    expect(setModalContentMock).toHaveBeenCalledTimes(1);
    expect(setModalContentMock).toHaveBeenCalledWith(<MovieDetail movieTitle={movie.name} />);
  });
});
