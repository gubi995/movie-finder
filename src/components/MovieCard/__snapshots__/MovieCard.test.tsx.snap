// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard component should render movie data 1`] = `
.emotion-0 {
  background-color: #fff;
  color: rgba(0, 0, 0, 0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 4px;
  box-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

@media (min-width: 768px) {
  .emotion-0 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

.emotion-1 {
  padding: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 2rem;
  gap: 1rem;
  -webkit-flex: 1 1;
  -ms-flex: 1 1;
  flex: 1 1;
}

.emotion-1:last-child {
  padding-bottom: 24px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  gap: 0.5rem;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.5rem;
}

.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
}

.emotion-4::-moz-focus-inner {
  border-style: none;
}

.emotion-4.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-4 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-5 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1.5rem;
  line-height: 1.334;
  letter-spacing: 0em;
  font-family: Montserrat;
}

.emotion-6 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  letter-spacing: 0.00938em;
  color: gray;
}

.emotion-7 {
  display: grid;
  place-items: center;
  padding: 0.5em;
  border-radius: 50%;
  width: 30px;
  height: 30px;
  font-weight: bold;
  background: linear-gradient(90deg, #00d2ff 0%, #3a47d5 100%);
  color: white;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5rem;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-9 {
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-size: 0.8125rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  height: 32px;
  color: rgba(0, 0, 0, 0.87);
  background-color: rgba(0, 0, 0, 0.08);
  border-radius: 16px;
  white-space: nowrap;
  -webkit-transition: background-color 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  cursor: default;
  outline: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  border: 0;
  padding: 0;
  vertical-align: middle;
  box-sizing: border-box;
}

.emotion-9.Mui-disabled {
  opacity: 0.38;
  pointer-events: none;
}

.emotion-9 .MuiChip-avatar {
  margin-left: 5px;
  margin-right: -6px;
  width: 24px;
  height: 24px;
  color: #616161;
  font-size: 0.75rem;
}

.emotion-9 .MuiChip-avatarColorPrimary {
  color: #fff;
  background-color: #1565c0;
}

.emotion-9 .MuiChip-avatarColorSecondary {
  color: #fff;
  background-color: #7b1fa2;
}

.emotion-9 .MuiChip-avatarSmall {
  margin-left: 4px;
  margin-right: -4px;
  width: 18px;
  height: 18px;
  font-size: 0.625rem;
}

.emotion-9 .MuiChip-icon {
  color: #616161;
  margin-left: 5px;
  margin-right: -6px;
}

.emotion-9 .MuiChip-deleteIcon {
  -webkit-tap-highlight-color: transparent;
  color: rgba(0, 0, 0, 0.26);
  font-size: 22px;
  cursor: pointer;
  margin: 0 5px 0 -6px;
}

.emotion-9 .MuiChip-deleteIcon:hover {
  color: rgba(0, 0, 0, 0.4);
}

.emotion-10 {
  overflow: hidden;
  text-overflow: ellipsis;
  padding-left: 12px;
  padding-right: 12px;
  white-space: nowrap;
}

.emotion-11 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  letter-spacing: 0.00938em;
}

<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root emotion-0"
    >
      <div
        class="MuiCardContent-root emotion-1"
      >
        <div
          class="emotion-2"
        >
          <div
            class="emotion-3"
          >
            <button
              class="MuiButtonBase-root emotion-4"
              tabindex="0"
              type="button"
            >
              <h5
                class="MuiTypography-root MuiTypography-h5 emotion-5"
              >
                Fight Club
              </h5>
            </button>
            <p
              class="MuiTypography-root MuiTypography-body1 emotion-6"
            >
              15/10/1999
            </p>
          </div>
          <div
            class="emotion-7"
          >
            8.4
          </div>
        </div>
        <div
          class="emotion-8"
        >
          <div
            class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault emotion-9"
          >
            <span
              class="MuiChip-label MuiChip-labelMedium emotion-10"
            >
              Drama
            </span>
          </div>
        </div>
        <p
          class="MuiTypography-root MuiTypography-body1 emotion-11"
        >
          A ticking-time-bomb insomniac and a slippery soap salesman channel primal male aggression into a shocking new form of therapy. Their concept catches on, with underground "fight clubs" forming in every town, until an eccentric gets in the way and ignites an out-of-control spiral toward oblivion.
        </p>
      </div>
    </div>
  </div>
</body>
`;
